### ISSUE TITLE
Two issues with local near cache

### ISSUE THREAD

**Issue by lukasblu** [hazelcast/hazelcast#4671]
Hi,

while stress testing our application, we discovered two issues with using local near caches:

1. Even on a single node, the local near cache is not synchronously invalidated after the map is modified.

2. Also on a single node, it may happen that near cache invalidations are lost.

To 1:
I understand that you do not give full consistency guarantees when using a near cache, however I kind of expected that at least the local node always sees its latest changes.
While adding full consistency guarantees on a distributed near cache might be very expensive (in terms of synchronization and operation latency), providing such guarantees on the local node only should be much cheaper.

To 2:
This is certainly a bug.

Please find below a test case which shows both issues.

If using a ConcurrentHashMap, the test always succeeds. If using a distributed Hazelcast map with local near cache enabled, the test either fails with
  "Assertion violated."
which means that the near cache showed old data, or the test fails with
  "Near cache did *not* become consistent."
which means that even after waiting for some time, the near cache did not show the latest data - hence near cache invalidations were lost.

It would be great to hear your opinion on 1 and a fix for 2 would be amazing! :-)

Thanks and best,
Lukas

```java
package com.nm.test.hazelcast.nearcache;

import com.hazelcast.config.*;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.nm.test.hazelcast.TestHazelcast;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import junit.framework.TestCase;

/**
 * A test to ensure the near cache eventually reflects the data from the map.
 */
public class TestLocalNearCache6 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestLocalNearCache6.class);

	private static final int numGetters = 7;

	private static final String mapName = "testMap" + TestLocalNearCache6.class.getSimpleName();

	private static final String key = "key123";

	private AtomicInteger valuePut = new AtomicInteger(0);

	private AtomicBoolean stop = new AtomicBoolean(false);

	private AtomicBoolean failed = new AtomicBoolean(false);

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testConsistentNearCache() throws Exception {

		// create concurrent hash map
		ConcurrentMap<String, String> map = new ConcurrentHashMap<String, String>();

		// run test
		testInternal(map);

		// fail if failed
		if (failed.get()) {
			fail("Assertion violated.");
		}
	}

	public void testConsistentNearCacheHz() throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// configure near cache
		NearCacheConfig nearCacheConfig = new NearCacheConfig();
		nearCacheConfig.setCacheLocalEntries(true); // this enables caching of local entries
		nearCacheConfig.setInMemoryFormat(InMemoryFormat.OBJECT);

		// enable near cache
		MapConfig mapConfig = config.getMapConfig(mapName);
		mapConfig.setNearCacheConfig(nearCacheConfig);

		// create Hazelcast instance
		HazelcastInstance hcInstance = Hazelcast.newHazelcastInstance(config);

		// get hazelcast map
		ConcurrentMap<String, String> map = hcInstance.getMap(mapName);

		// run test
		testInternal(map);

		// test eventually consistent
		Thread.sleep(1000);
		int valuePutLast = valuePut.get();
		String valueMapStr = map.get(key);
		int valueMap = Integer.parseInt(valueMapStr);

		// stop hazelcast
		hcInstance.getLifecycleService().terminate();

		// test eventually consistent
		if (valueMap < valuePutLast) {
			String msg = "Near cache did *not* become consistent. (valueMap = " + valueMap + ", valuePut = " + valuePutLast + ")";
			logger.warn(msg);
			fail(msg);
		}

		// fail if failed
		if (failed.get()) {
			fail("Assertion violated.");
		}
	}

	private void testInternal(ConcurrentMap<String, String> map) throws Exception {

		// start 1 putter thread
		Thread threadPut = new Thread(new PutRunnable(map), "put");
		threadPut.start();
		Thread.sleep(1000);

		// start numGetters getter threads
		List<Thread> threads = new ArrayList<Thread>();
		for (int i = 0; i < numGetters; i++) {
			Thread thread = new Thread(new GetRunnable(map), "get" + i);
			threads.add(thread);
		}
		for (Thread thread : threads) {
			thread.start();
		}

		// stop after 10s
		int i = 0;
		while (!stop.get() && i++ < 10) {
			Thread.sleep(1000);
		}
		if (i == 20) {
			logger.info("Problem did not occur within 10s.");
		}
		stop.set(true);
		threadPut.join();
		for (Thread thread : threads) {
			thread.join();
		}
	}

	private class PutRunnable implements Runnable {

		private final ConcurrentMap<String, String> map;

		public PutRunnable(ConcurrentMap<String, String> map) {
			this.map = map;
		}

		@Override
		public void run() {
			logger.info(Thread.currentThread().getName() + " started.");
			int i = 0;
			while (!stop.get()) {

				// put new value and update last state
				// note: the value in the map/near cache is *always* larger or equal to valuePut
				// assert: valueMap >= valuePut
				i++;
				map.put(key, String.valueOf(i));
				valuePut.set(i);

				// sleep
				try {
					Thread.sleep(50);
				} catch (InterruptedException e) {
					logger.warn("Interrupted: " + e.getMessage());
				}
			}
			logger.info(Thread.currentThread().getName() + " performed " + i + " operations.");
		}

	}

	private class GetRunnable implements Runnable {

		private final ConcurrentMap<String, String> map;

		public GetRunnable(ConcurrentMap<String, String> map) {
			this.map = map;
		}

		@Override
		public void run() {
			logger.info(Thread.currentThread().getName() + " started.");
			int n = 0;
			while (!stop.get()) {

				// compare value from near cache with last put value
				int valuePutLast = valuePut.get();
				String valueMapStr = map.get(key);
				int valueMap = Integer.parseInt(valueMapStr);
				n++;

				// ensure assertion holds
				if (valueMap < valuePutLast) {
					logger.warn("Assertion violated! (valueMap = " + valueMap + ", valuePut = " + valuePutLast + ")");
					failed.set(true);
					stop.set(true);
				}
			}
			logger.info(Thread.currentThread().getName() + " performed " + n + " operations.");
		}

	}

}
```

**Comment by lukasblu** [comment]
Ups, just discovered a small typo in the test - however, it has no influence on its outcome (only the success messages is never printed).

**Comment by lukasblu** [comment]
Ps. I tested with version 3.4.1 and also with version 3.3.4 - both issues occur in both versions.

Pps. as this is a stress test, the issues does not always appear... it generally helps to make sure your system is under full load (increase var numGetters to something above your core count) and decreasing the sleep time in the putter thread also helps.

**Comment by metanet** [comment]
Near cache invalidations are not lost. You do not wait enough to let it become consistent (eventually). 

 ```Thread.sleep(1000L); ```

If you add this line at the beginning of your while loop in GetRunnable, you will see that HZ test will pass as well. 

Besides, I am not sure if it intentional or not but you do not get every value you put into the map. 

 ```
// compare value from near cache with last put value
int valuePutLast = valuePut.get();
String valueMapStr = map.get(key);
int valueMap = Integer.parseInt(valueMapStr);
n++; 
``` 

This code does not get every value that has been put into the map. 
For instance, in your PutRunnable you can put 1,2,3,4,5 and you may only test 5 and skip 1,2,3,4.

**Comment by lukasblu** [comment]
Hi @metanet ,

thanks for answers.

Can you show me where exactly you put the Thread.sleep(1000L)? Because inside the while loop in the GetRunnable it really should not affect this. Note, however, that if you sleep 1s in every loop in the GetRunnable, then the problem described in this issue naturally becomes *much* less likely - I think the issue occurs if *very shortly* after the IMap.put() you call IMap.get(). I try to enforce such a situation with this test.

The test makes no effort at all to retrieve all the values put into the map. A GetRunnable may skip lots of values - that is fine. All that matters to me is that a GetRunnable *never* sees at newer value in valuePutLast compared to what it gets from the map. This is the invariation which must never be broken.

Best,
Lukas

**Comment by metanet** [comment]
Hi @lukasblu 

I put that line after  while (!stop.get()) in GetRunnable.

You are right, when I increase the thread count and decrease the sleep times, your assumption gets broken. 

Thanks for your explanation.

**Comment by lukasblu** [comment]
Hej, I wrote another test which more clearly shows that near cache invalidations can be lost - and which also proves this by flushing the near cache and checking again.

Also, the test is a bit easier to understand compared to the first one.

However, it still is a stress test and therefore does not fail all the times. If it does not fail for you, play with the sleep time in the putter thread, e.g. decrease to 10ms or increase to 25ms.

Cheers,
Lukas

```java
package com.nm.test.hazelcast.nearcache;

import com.hazelcast.config.*;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.instance.HazelcastInstanceImpl;
import com.hazelcast.instance.HazelcastInstanceProxy;
import com.hazelcast.map.impl.MapService;
import com.hazelcast.map.impl.MapServiceContext;
import com.hazelcast.map.impl.NearCacheProvider;
import com.hazelcast.spi.impl.NodeEngineImpl;
import com.nm.test.hazelcast.TestHazelcast;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import junit.framework.TestCase;

/**
 * A test to ensure no lost invalidations on the near cache.
 */
public class TestLocalNearCache7 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestLocalNearCache7.class);

	private static final int numGetters = 7;

	private static final int maxRuntime = 30;

	private static final String mapName = "testMap" + TestLocalNearCache7.class.getSimpleName();

	private static final String key = "key123";

	private ConcurrentMap<String, String> map;

	private AtomicInteger valuePut = new AtomicInteger(0);

	private AtomicBoolean stop = new AtomicBoolean(false);

	private AtomicBoolean failed = new AtomicBoolean(false);

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testNoLostInvalidations() throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// we use a single node only - do not search for others
		config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);

		// configure near cache
		NearCacheConfig nearCacheConfig = new NearCacheConfig();
		nearCacheConfig.setCacheLocalEntries(true); // this enables caching of local entries
		nearCacheConfig.setInMemoryFormat(InMemoryFormat.OBJECT);

		// enable near cache
		MapConfig mapConfig = config.getMapConfig(mapName);
		mapConfig.setNearCacheConfig(nearCacheConfig);

		// create Hazelcast instance
		HazelcastInstance hcInstance = Hazelcast.newHazelcastInstance(config);

		// get hazelcast map
		map = hcInstance.getMap(mapName);

		// run test
		runTestInternal();

		// test eventually consistent
		Thread.sleep(2000);
		int valuePutLast = valuePut.get();
		String valueMapStr = map.get(key);
		int valueMap = Integer.parseInt(valueMapStr);

		// fail if not eventually consistent
		String msg = null;
		if (valueMap < valuePutLast) {
			msg = "Near cache did *not* become consistent. (valueMap = " + valueMap + ", valuePut = " + valuePutLast + ").";

			// flush near cache and re-fetch value
			flushNearCache(hcInstance);
			String valueMap2Str = map.get(key);
			int valueMap2 = Integer.parseInt(valueMap2Str);

			// test again
			if (valueMap2 < valuePutLast) {
				msg += " Unexpected inconsistency! (valueMap2 = " + valueMap2 + ", valuePut = " + valuePutLast + ").";
			} else {
				msg += " Flushing the near cache cleared the inconsistency. (valueMap2 = " + valueMap2 + ", valuePut = " + valuePutLast + ").";
			}
		}

		// stop hazelcast
		hcInstance.getLifecycleService().terminate();

		// fail after stopping hazelcast instance
		if (msg != null) {
			logger.warn(msg);
			fail(msg);
		}
	}

	/**
	 * Flush near cache.
	 * <p>
	 * Warning: this uses Hazelcast internals which might change from one version to the other.
	 */
	private void flushNearCache(HazelcastInstance hcInstance) throws Exception {

		// get instance proxy
		HazelcastInstanceProxy hcInstanceProxy = (HazelcastInstanceProxy) hcInstance;

		// get internal implementation (using reflection)
		Field originalField = HazelcastInstanceProxy.class.getDeclaredField("original");
		originalField.setAccessible(true);
		HazelcastInstanceImpl hcImpl = (HazelcastInstanceImpl) originalField.get(hcInstanceProxy);

		// get node engine
		NodeEngineImpl nodeEngineImpl = hcImpl.node.nodeEngine;

		// get map service
		MapService mapService = nodeEngineImpl.getService(MapService.SERVICE_NAME);

		// clear near cache
		MapServiceContext mapServiceContext = mapService.getMapServiceContext();
		NearCacheProvider nearCacheProvider = mapServiceContext.getNearCacheProvider();
		nearCacheProvider.clearNearCache(mapName);
	}

	private void runTestInternal() throws Exception {

		// start 1 putter thread (put0)
		Thread threadPut = new Thread(new PutRunnable(), "put0");
		threadPut.start();

		// wait for putter thread to start before starting getter threads
		Thread.sleep(300);

		// start numGetters getter threads (get0-numGetters)
		List<Thread> threads = new ArrayList<Thread>();
		for (int i = 0; i < numGetters; i++) {
			Thread thread = new Thread(new GetRunnable(), "get" + i);
			threads.add(thread);
		}
		for (Thread thread : threads) {
			thread.start();
		}

		// stop after maxRuntime seconds
		int i = 0;
		while (!stop.get() && i++ < maxRuntime) {
			Thread.sleep(1000);
		}
		if (!stop.get()) {
			logger.info("Problem did not occur within " + maxRuntime + "s.");
		}
		stop.set(true);
		threadPut.join();
		for (Thread thread : threads) {
			thread.join();
		}
	}

	private class PutRunnable implements Runnable {

		@Override
		public void run() {
			logger.info(Thread.currentThread().getName() + " started.");
			int i = 0;
			while (!stop.get()) {
				i++;

				// put new value and update last state
				// note: the value in the map/near cache is *always* larger or equal to valuePut
				// assertion: valueMap >= valuePut
				map.put(key, String.valueOf(i));
				valuePut.set(i);

				// sleep
				// (this is needed such that we only stop the test after a near cache invalidation was permanently lost)
				try {
					Thread.sleep(20);
				} catch (InterruptedException e) {
					logger.warn("Interrupted: " + e.getMessage());
				}

				// check if we see our last update
				String valueMapStr = map.get(key);
				int valueMap = Integer.parseInt(valueMapStr);
				if (valueMap != i) {
					logger.warn("Assertion violated! (valueMap = " + valueMap + ", i = " + i + ")");
					failed.set(true);
					stop.set(true);
				}
			}
			logger.info(Thread.currentThread().getName() + " performed " + i + " operations.");
		}

	}

	private class GetRunnable implements Runnable {

		@Override
		public void run() {
			logger.info(Thread.currentThread().getName() + " started.");
			int n = 0;
			while (!stop.get()) {
				n++;

				// blindly get the value - to trigger issue - and
				// parse the value - to get some CPU load
				String valueMapStr = map.get(key);
				Integer.parseInt(valueMapStr);
			}
			logger.info(Thread.currentThread().getName() + " performed " + n + " operations.");
		}

	}

}
```

**Comment by lukasblu** [comment]
Hi again, I just now found a way to make the test fail *all the times*: Instead of sleeping between the IMap.put() and IMap.get(), I now sleep only after I have already discovered an inconsistency.

Please replace the implementation of the PutRunnable with the following code:

```java
	private class PutRunnable implements Runnable {

		@Override
		public void run() {
			logger.info(Thread.currentThread().getName() + " started.");
			int i = 0;
			while (!stop.get()) {
				i++;

				// put new value and update last state
				// note: the value in the map/near cache is *always* larger or equal to valuePut
				// assertion: valueMap >= valuePut
				map.put(key, String.valueOf(i));
				valuePut.set(i);

				// check if we see our last update
				String valueMapStr = map.get(key);
				int valueMap = Integer.parseInt(valueMapStr);
				if (valueMap != i) {
					logger.warn("Assertion violated! (valueMap = " + valueMap + ", i = " + i + ")");

					// sleep to ensure near cache invalidation is really lost
					try {
						Thread.sleep(100);
					} catch (InterruptedException e) {
						logger.warn("Interrupted: " + e.getMessage());
					}

					// test again and stop if really lost
					valueMapStr = map.get(key);
					valueMap = Integer.parseInt(valueMapStr);
					if (valueMap != i) {
						logger.warn("Near cache invalidation lost! (valueMap = " + valueMap + ", i = " + i + ")");
						failed.set(true);
						stop.set(true);
					}
				}
			}
			logger.info(Thread.currentThread().getName() + " performed " + i + " operations.");
		}

	}
```

**Comment by serkan-ozal** [comment]
@lukasblu

I think, the problematic flow can be like this (based on first version of your test code):

**Note:** `Putter`, `Getter-1` and `Getter-2` runs on different threads.

**1.** `Putter` puts value `1` and now `valuePut` is `1`
**2.** `Getter-1` tries to get value 
&nbsp;&nbsp;&nbsp;&nbsp;**2.1.** Since near-cache is empty, request from map via operiation
&nbsp;&nbsp;&nbsp;&nbsp;**2.2.** Get value `1` from map 
&nbsp;&nbsp;&nbsp;&nbsp;**2.3.** Return value `1` to caller and start an asynchronous execution callback (`Callback-1`) to put value `1` into near-cache
**3.** `Putter` puts value `2`  and now `valuePut` is `2`
**4.** `Callback-1` puts value `1` to near-cache
**5.** `Getter-2` tries to get value 
&nbsp;&nbsp;&nbsp;&nbsp;**5.1.** Since the value is `1` in near-cache (put by `Callback-1` as async), return value `1` to caller
**6.** Since value got by `Getter-2` is `1` and `valuePut` is `2`, test fails

**Comment by serkan-ozal** [comment]
By the way, I am not sure about is this really a bug or not from the perspective of async save to near-cache. But maybe we can support synch save configuration (by default this configuration can be disabled and user can enable this) to near-cache to handle this problem but it is obvious that sync save to near-cache will have performance trade-off from the caller side.

**Comment by serkan-ozal** [comment]
In addition, maybe while invalidation near-cache record on put to map, we may also cancel all waiting callbacks (not executed yet by executor service and will save to near-cache) for that key (possibly by holding key-callback(s) mappings).

**Comment by lukasblu** [comment]
Hi @serkan-ozal , thanks for your feedback.

I have just one comment on point 2.3: Where is this Callback-1 executed? I was looking at the code in MapProxyImpl and MapProxySupport, but could not find it. Furthermore, is such a callback really needed for the local node? If you look at the implementation of MapProxySupport.getInternal(), you can see that the value returned by the map operation is *synchronously* put into the near cache.

Thanks and best,
Lukas

**Comment by serkan-ozal** [comment]
Hi @lukasblu,

yes you are right. I had looked `getAsync` method by mistake. Currently it really seems like a bug. Let me look in detail again.

**Comment by serkan-ozal** [comment]
Hi @lukasblu,

What about this flow for sync get:

**Note:** `Putter`, `Getter-1` runs on different threads.

**1.** `Putter` puts value `1`
&nbsp;&nbsp;&nbsp;&nbsp;**1.1.** Invalide near-cache record (no invalidation occures since there is no near-cache record yet) 
&nbsp;&nbsp;&nbsp;&nbsp;**1.2.** Now `valuePut` is `1`
**2.** `Putter` thread is suspended and `Getter-1` thread is scheduled by OS
**3.** `Getter-1` tries to get value 
&nbsp;&nbsp;&nbsp;&nbsp;**3.1.** Since near-cache is empty, request from map via operiation
&nbsp;&nbsp;&nbsp;&nbsp;**3.2.** Get value `1` from map 
&nbsp;&nbsp;&nbsp;&nbsp;**3.3.** `Getter-1` thread is suspended (at here https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java#L231) and `Putter` thread is scheduled by OS
**4.** `Putter` puts value `2` 
&nbsp;&nbsp;&nbsp;&nbsp;**4.1.** Invalide near-cache record (no invalidation occures since `Getter-1` has not put value `1` to near-cache yet) 
&nbsp;&nbsp;&nbsp;&nbsp;**4.2.** Now `valuePut` is `2`
&nbsp;&nbsp;&nbsp;&nbsp;**4.3.** `Putter` thread is suspended and `Getter-1` thread is scheduled (from here https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java#L231) again by OS
**5.** `Getter-1` puts value `1` (taken from map via operation before value `2` is put at step `3.2.`) to near cache
**6.** `Getter-1` returns value `1` to caller
**7.** Since value got by `Getter-1` is `1` and `valuePut` is `2`, test fails

So the problem is about non-atomic put/get operation on IMap and Near-Cache

**Comment by lukasblu** [comment]
Hi @serkan-ozal ,

yes, this sounds like a possible explanation.

It also explains, why sometimes a near cache value remains at the old value: Not only the local, synchronous invalidation of the near cache is missed, but also the invalidate operation sent out after the map is updated. If this also happens before Getter-1 inserts its previously fetched value into the near cache, then nobody will invalidate that value afterwards.

Now, do you have a suggestion how to fix this? :-)

I see several alternatives:

(a) the map entry could be locked before updating the near cache:
1. lock map entry before executing the GetOperation here: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java#L225 (you would only do that if near cache is enabled)
2. execute GetOperation and put value into near cache
3. release lock
=> this approach potentially requires two additional network operations.

(b) the near cache entry for the given map entry could be locked before updating the near cache:
1. lock near cache entry (also if it does not exist)
2. execute GetOperation and put value into near cache
3. unlock near cache entry
To implement this approach, a ConcurrentHashMap which contains keys for the currently locked near cache entries could be used.
=> this approach requires *no* additional network operations.

Alternative (b) currently sounds quite promising to me: No additional network operations, Little memory use, quite easy to implement. The only difficulty I see at the moment is how to ensure that locked near cache entries do not block invalidate operations forever. However, maybe this is *not* a problem, since every remote operation in Hazelcast already has a timeout and so updating the near cache will never wait longer than this timeout.

What do you think?

Cheers,
Lukas

**Comment by lukasblu** [comment]
Short status update:

I did not try to implement alternative (a), but I did try to implement alternative (b). Unfortunately, with alternative (b) I ran into deadlocks between the putter and/or getter threads and the Hazelcast partition thread. The problem was that the Hazelcast partition thread queues pending operations and if you are unlucky, then the operation queued by the thread which currently has the lock is *not* the next one to be executed - so you run into a deadlock.

The good news: I implemented another alternative (c) which goes like this: On the near cache I keep a invalidate operation count which is increased upon every call to NearCache.invalidate(). Now before updating the near cache with a new value fetched by an execution of a GetOperation, I remember the current invalidate operation count. This remembered invalidate operation count is then passed to the putNearCache() function and there, I only update the near cache if there was no new invalidate operation in between.

I will create a preliminary pull request and link it here for you to review...

Cheers,
Lukas

**Comment by serkan-ozal** [comment]
Hi @lukasblu,

I had though similar thing, but in that solution, operation count should be key based. Otherwise, you may ignore other key updates unnecessarily. Keeping key based operation count is not feasible.

**Comment by serkan-ozal** [comment]
Hi @lukasblu,

In addition, to make key based operation count feasible, I think, we can use segmented operation count as hash codes of keys (such as 256 operation count of 256 key segments). So in this case, we minimize the scope of unnecessary key update ignoring.

**Comment by lukasblu** [comment]
Hi @serkan-ozal ,

Thanks for your feedback. Key-based operation counts might be an option too, but I'm not (yet) fully convinced that my fix above really misses key updates. Could you detail in which situations such an updated would be missed?

Thanks,
Lukas

**Comment by serkan-ozal** [comment]
Hi @lukasblu,

As I understood from your approach, this flow can be problem.

`OPC:` Operation Counter

**1.** Put (`Key-1`, `Value-1.1`) (Set `OPC` to `1`)
**2.** Get `Key-1` (Get `OPC` as 1) and `Getter-1` thread is suspended before return to caller
**3.** Get `Key-2` (Get `OPC` as 1) and `Getter-2` thread is suspended before return to caller
**4.** Put (`Key-1`, `Value-1.2`) (Set `OPC` to `2`)
**5.** `Getter-1` thread is scheduled and since its `OPC` (`1`) is less than current `OPC` (`2`), its taken value is ignored on put to near-cache. This is expected behaviour since value of `Key-1` is updated and should not be added to near-cache.
**6.** `Getter-2` thread is scheduled and since its `OPC` (`1`) is less than current `OPC` (`2`), its taken value is ignored on put to near-cache. This is **not** expected behaviour since value of `Key-2` is **not** updated and taken value for `Key-2` should be added to near-cache. I accept that at next calls for `Key-2`, its value will be put into near-cache eventually. But I expect that it should put into near-cache at first call

Am I missing something in your idea ?

Thanks.

**Comment by lukasblu** [comment]
Hi @serkan-ozal ,

thanks again for your explanations.

True, in my implementation I keep a single invalidate count per NearCache. However, I also keep invalidation information *per key*. When updating the near cache I accept (or ignore) updates based on this per key invalidation information. See NearCache.put() in my implementation.

To avoid a memory leak, I attached this invalidation information to regular CacheRecord objects and therefore, if no longer needed, the invalidation information is removed from the near cache again.

Do you think this works?

Thanks and best,
Lukas

**Comment by serkan-ozal** [comment]
Hi @lukasblu,

OK, I had missed that part in your code.
But as I see from your PR, on invalidate we don't remove near-cache record.
Doesn't it cause to memory problem ? 
Is there any reason to hold near-cache record of invalidated keys ?

**Comment by lukasblu** [comment]
Hi @serkan-ozal ,

sure, it will keep those CacheRecord objects a bit longer in the near cache, but we need this information, otherwise we would not know that a specific key was invalidated. This is the per key invalidation information that I was talking about above.

Even though it keeps those CacheRecord objects around a bit longer, this will not result in a memory leak as long as you have some reasonable TTL, max size, or max idle configuration set. To support installations with no near cache eviction strategy at all, we could also simply evict *invalid* CacheRecord objects if they are older than a certain time - e.g. something a little above twice the operation timeout which you already have in Hazelcast.

Okay?

Best,
Lukas

**Comment by lukasblu** [comment]
Hi @serkan-ozal ,

one optimization concerning memory usage would be to set the CacheRecord.value to null on invalidate. This would probably release already most of the addionally kept memory.

Should I update my PR?

Cheers,
Lukas

**Comment by serkan-ozal** [comment]
Hi @lukasblu,

From my side, I am really not sure keeping records of invalidated entries even though we will clear them eventually. In case of highly put (invalidate) operations, it creates so many new objects (records), it may fill up TLAB (so Eden space) and they can be moved to Tenured space (through Survivor spaces) before we collect them.

And also beside high memory usage, there are more operations (internal CHM accesses), and I think, obviously it will affect near-cache performance.

In addition, I have another two candidates:

* Similar approach as you but with segmented operation counter
* Making near-cache eventually consistent by a background thread that updates near-cache items from original cache (not detailed yet) and this can be enabled with a flag on near cache configuration

**Comment by lukasblu** [comment]
Hi @serkan-ozal ,

thanks for your feedback.

For me it is absolutely fine if you want to implement another solution. However, I think the additional memory consumption is not too bad because of two reasons:

1. The near cache is often used for large objects (to avoid fetching them from remote nodes, or to safe deserialization cost). Therefore, the near cache anyway uses a lot of memory and the additional memory used to keep the invalidation information per key is small in comparison.

2. The invalidation information is kept per key. Therefore, as long as you work with existing map entries, the invalidation information is only kept until the map entry is fetched again.

I have updated my pull request #4710 to drop the invalidated value from the CacheRecord object and then tested it in a real life application - and, I did not see any increase in memory usage even though I had a working set of several thousand map entries which changed over time.

For me it is very important that this bug is fixed, because the problem - when it happens - is very nasty and hard to debug. Therefore, please either merge my pull request or implement your another solution.

Thanks and best,
Lukas

**Comment by mtopolnik** [comment]
I would just comment on the first point by lukasblu here:

>1. The near cache is often used for large objects

Solutions must be designed for the worst case, not for the typical case. Considering the impact of the solution with *very small* objects is mandatory.

**Comment by mtopolnik** [comment]
In reference to https://github.com/hazelcast/hazelcast/issues/4671#issuecomment-76045979, unless:
1. the task doing the callback cancellation is on the same queue as the callbacks it is about to cancel;
2. the queue is being consumed by a single thread;
3. the cancelling task is ahead of the callbacks in that queue,
there can be no guarantee that the callbacks will actually be cancelled before starting.

**Comment by lukasblu** [comment]
Hi,

just a short update on the status of this issue: I run my tests using 3.5 and this problem still exists.

I will now try the propsed fix by @ahmetmircik here:
https://github.com/hazelcast/hazelcast/compare/master...ahmetmircik:fix/3.5/nearcacheStalenessMaybeFix?expand=1

Best,
Lukas

**Comment by mtopolnik** [comment]
@lukasblu Don't; we're already working on it and Ahmet has a much better, deterministic approach where correct ordering of cache operations is enforced by the operation thread. This work did not make it to the 3.5.0 release due to time constraints. If you do go and experiment with the above WIP, may I warn you that `getSlot()` there is probably broken because it is using a bitmask approach to effect a modulo operation with a non-power-of-two modulus (partition count).

**Comment by lukasblu** [comment]
Hi @mtopolnik ,

oh, I just saw your comment now! :-) So, I did test the "possible fix" of @ahmetmircik and it looks promising, but you are right, it does _not_ fully fix the problem.

What I experienced was, that no near cache invalidations were actually lost, but the put thread in my TestLocalNearCache7 test is still seeing assertion violations - meaning that the put thread does not always see its latest updates.

It would be great if you could test your new approach against the TestLocalNearCache7 test as well and make sure those assertions are always kept.

If you want, I can also help to test the new implementation once it is in a more or less stable state.

Thanks and best,
Lukas

**Comment by mtopolnik** [comment]
@lukasblu I have imported your test code as an official Hazelcast test case so rest assured its assertions will be satisfied :)

**Comment by lukasblu** [comment]
Hi @mtopolnik ,

that sounds great!

My only concern left is, that I'm not fully sure if my test is _strict_ enough. If you take a look at my last code snipped I added to this issue (the improved PutRunnable) you will see the following line:

```java
logger.warn("Assertion violated! (valueMap = " + valueMap + ", i = " + i + ")");
```

I print this warning if the assertion is violated. But, instead of failing the test immediately, I test again to see if the near cache becomes eventually consistent. If that is the case, I continue the test and may successfully complete it.

Even though the test then passes, it still violates my assertions - meaning that the local near cache is not fully consistent, but only eventually consistent.

What do you think? Should the local near cache be fully consistent, or is it enough to have it eventually consistent? For programming it would make it easier to have the local near cache fully consistent.

Thanks and best,
Lukas

**Comment by mtopolnik** [comment]
@lukasblu 

Yes, I noticed this subtlety... I had my own fix (too complex to go to production) which never produced any of those warnings so my view is that those should be promoted to full-blown assertion failures.

**Comment by lukasblu** [comment]
Hi @mtopolnik 

nice, so we agree :-)

Here is an updated version of TestLocalNearCache7 which performs two tests - one to ensure eventual consistency and one to be strict on all assertions:

```java
package com.nm.test.hazelcast.nearcache;

import com.hazelcast.config.*;
import com.hazelcast.core.Hazelcast;
import com.hazelcast.core.HazelcastInstance;
import com.hazelcast.instance.HazelcastInstanceImpl;
import com.hazelcast.instance.HazelcastInstanceProxy;
import com.hazelcast.map.impl.MapService;
import com.hazelcast.map.impl.MapServiceContext;
import com.hazelcast.map.impl.NearCacheProvider;
import com.hazelcast.spi.impl.NodeEngineImpl;
import com.nm.test.hazelcast.TestHazelcast;
import org.apache.log4j.BasicConfigurator;
import org.apache.log4j.Logger;
import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ConcurrentMap;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import junit.framework.TestCase;

/**
 * A test to ensure no lost invalidations on the near cache.
 * <p>
 * Issue: https://github.com/hazelcast/hazelcast/issues/4671
 */
public class TestLocalNearCache7 extends TestCase {

	private static final Logger logger = Logger.getLogger(TestLocalNearCache7.class);

	private static final int numGetters = 7;

	private static final int maxRuntime = 30;

	private static final String mapName = "testMap" + TestLocalNearCache7.class.getSimpleName();

	private static final String key = "key123";

	private ConcurrentMap<String, String> map;

	private AtomicInteger valuePut = new AtomicInteger(0);

	private AtomicBoolean stop = new AtomicBoolean(false);

	private AtomicInteger assertionViolationCount = new AtomicInteger(0);

	private AtomicBoolean failed = new AtomicBoolean(false);

	@Override
	protected void setUp() throws Exception {

		// configure logging
		if (!TestHazelcast.loggingInitialized) {
			TestHazelcast.loggingInitialized = true;
			BasicConfigurator.configure();
		}
	}

	public void testNoLostInvalidationsEventually() throws Exception {
		assertionViolationCount.set(0);
		testNoLostInvalidations(false);
	}

	public void testNoLostInvalidationsStrict() throws Exception {
		assertionViolationCount.set(0);
		testNoLostInvalidations(true);
	}

	private void testNoLostInvalidations(boolean strict) throws Exception {

		// create hazelcast config
		Config config = new XmlConfigBuilder().build();
		config.setProperty("hazelcast.logging.type", "log4j");

		// we use a single node only - do not search for others
		config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);

		// configure near cache
		NearCacheConfig nearCacheConfig = new NearCacheConfig();
		nearCacheConfig.setCacheLocalEntries(true); // this enables caching of local entries
		nearCacheConfig.setInMemoryFormat(InMemoryFormat.OBJECT);

		// enable near cache
		MapConfig mapConfig = config.getMapConfig(mapName);
		mapConfig.setNearCacheConfig(nearCacheConfig);

		// create Hazelcast instance
		HazelcastInstance hcInstance = Hazelcast.newHazelcastInstance(config);

		// get hazelcast map
		map = hcInstance.getMap(mapName);

		// run test
		runTestInternal();

		// test eventually consistent
		Thread.sleep(2000);
		int valuePutLast = valuePut.get();
		String valueMapStr = map.get(key);
		int valueMap = Integer.parseInt(valueMapStr);

		// fail if not eventually consistent
		String msg = null;
		if (valueMap < valuePutLast) {
			msg = "Near cache did *not* become consistent. (valueMap = " + valueMap + ", valuePut = " + valuePutLast + ").";

			// flush near cache and re-fetch value
			flushNearCache(hcInstance);
			String valueMap2Str = map.get(key);
			int valueMap2 = Integer.parseInt(valueMap2Str);

			// test again
			if (valueMap2 < valuePutLast) {
				msg += " Unexpected inconsistency! (valueMap2 = " + valueMap2 + ", valuePut = " + valuePutLast + ").";
			} else {
				msg += " Flushing the near cache cleared the inconsistency. (valueMap2 = " + valueMap2 + ", valuePut = " + valuePutLast + ").";
			}
		}

		// stop hazelcast
		hcInstance.getLifecycleService().terminate();

		// fail after stopping hazelcast instance
		if (msg != null) {
			logger.warn(msg);
			fail(msg);
		}

		// fail if strict is required and assertion was violated
		if (strict && assertionViolationCount.get() > 0) {
			msg = "Assertion violated " + assertionViolationCount.get() + " times.";
			logger.warn(msg);
			fail(msg);
		}
	}

	/**
	 * Flush near cache.
	 * <p>
	 * Warning: this uses Hazelcast internals which might change from one version to the other.
	 */
	private void flushNearCache(HazelcastInstance hcInstance) throws Exception {

		// get instance proxy
		HazelcastInstanceProxy hcInstanceProxy = (HazelcastInstanceProxy) hcInstance;

		// get internal implementation (using reflection)
		Field originalField = HazelcastInstanceProxy.class.getDeclaredField("original");
		originalField.setAccessible(true);
		HazelcastInstanceImpl hcImpl = (HazelcastInstanceImpl) originalField.get(hcInstanceProxy);

		// get node engine
		NodeEngineImpl nodeEngineImpl = hcImpl.node.nodeEngine;

		// get map service
		MapService mapService = nodeEngineImpl.getService(MapService.SERVICE_NAME);

		// clear near cache
		MapServiceContext mapServiceContext = mapService.getMapServiceContext();
		NearCacheProvider nearCacheProvider = mapServiceContext.getNearCacheProvider();
		nearCacheProvider.clearNearCache(mapName);
	}

	private void runTestInternal() throws Exception {

		// start 1 putter thread (put0)
		Thread threadPut = new Thread(new PutRunnable(), "put0");
		threadPut.start();

		// wait for putter thread to start before starting getter threads
		Thread.sleep(300);

		// start numGetters getter threads (get0-numGetters)
		List<Thread> threads = new ArrayList<Thread>();
		for (int i = 0; i < numGetters; i++) {
			Thread thread = new Thread(new GetRunnable(), "get" + i);
			threads.add(thread);
		}
		for (Thread thread : threads) {
			thread.start();
		}

		// stop after maxRuntime seconds
		int i = 0;
		while (!stop.get() && i++ < maxRuntime) {
			Thread.sleep(1000);
		}
		if (!stop.get()) {
			logger.info("Problem did not occur within " + maxRuntime + "s.");
		}
		stop.set(true);
		threadPut.join();
		for (Thread thread : threads) {
			thread.join();
		}
	}

	private class PutRunnable implements Runnable {

		@Override
		public void run() {
			logger.info(Thread.currentThread().getName() + " started.");
			int i = 0;
			while (!stop.get()) {
				i++;

				// put new value and update last state
				// note: the value in the map/near cache is *always* larger or equal to valuePut
				// assertion: valueMap >= valuePut
				map.put(key, String.valueOf(i));
				valuePut.set(i);

				// check if we see our last update
				String valueMapStr = map.get(key);
				int valueMap = Integer.parseInt(valueMapStr);
				if (valueMap != i) {
					assertionViolationCount.incrementAndGet();
					logger.warn("Assertion violated! (valueMap = " + valueMap + ", i = " + i + ")");

					// sleep to ensure near cache invalidation is really lost
					try {
						Thread.sleep(100);
					} catch (InterruptedException e) {
						logger.warn("Interrupted: " + e.getMessage());
					}

					// test again and stop if really lost
					valueMapStr = map.get(key);
					valueMap = Integer.parseInt(valueMapStr);
					if (valueMap != i) {
						logger.warn("Near cache invalidation lost! (valueMap = " + valueMap + ", i = " + i + ")");
						failed.set(true);
						stop.set(true);
					}
				}
			}
			logger.info(Thread.currentThread().getName() + " performed " + i + " operations.");
		}

	}

	private class GetRunnable implements Runnable {

		@Override
		public void run() {
			logger.info(Thread.currentThread().getName() + " started.");
			int n = 0;
			while (!stop.get()) {
				n++;

				// blindly get the value - to trigger issue - and
				// parse the value - to get some CPU load
				String valueMapStr = map.get(key);
				Integer.parseInt(valueMapStr);
			}
			logger.info(Thread.currentThread().getName() + " performed " + n + " operations.");
		}

	}

}
```

Thanks and best,
Lukas

**Comment by lukasblu** [comment]
Hi,

how is life? and how do we continue with this issue? In 3.5.1 there is still no fix.

The PR by @ahmetmircik #5550 got closed without merging.
The previous PR by @mtopolnik #5459 also got closed without merging.
My PR #4710 never got merged either
and there was an alternative idea from @ahmetmircik here
https://github.com/hazelcast/hazelcast/compare/master...ahmetmircik:fix/3.5/nearcacheStalenessMaybeFix?expand=1

Thanks and best,
Lukas

**Comment by ahmetmircik** [comment]
Hi @lukasblu, currently some solutions exist around but still we are not happy with any one of them. Lets continue to keep an eye on it.

**Comment by lukasblu** [comment]
Hi @ahmetmircik ,

I just wanted to ask if there is any progress on this issue.

For me it is very important to have no inconsistencies in the near cache and we are therefore still using my patch. At the moment we are thinking about upgrading to 3.5.x, but without a fix for this issue, I would have to re-apply my patch again... therefore, it would be good to know if you plan to come up with a fix in 3.5.3.

Thanks and best,
Lukas
